### PR TITLE
C++: Remove indirect -> direct taint-flow

### PR DESCRIPTION
**Pull request by MathiasVP**
The gist of this PR is super simple (it's the [first commit](c9c0243c50f34cbf7a5b5cda56bc2af75e0d7d6e)) but it sent me on a wild ride of bugfixing. So now I invite you, my dear reviewer, to join me on this wild ride!

Please review this PR on a commit-by-commit basis.

### CODE REVIEW

File: cpp/ql/test/TestUtilities/dataflow/FlowTestCommon.qll
Comment (jketema) [inline cpp/ql/test/TestUtilities/dataflow/FlowTestCommon.qll:32]: Do I read this correctly as counting the number of (line, column)-pairs?

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:1005]: The commit says "Fix a bug in Expr <-> Node mapping.". What is the bug exactly?

File: cpp/ql/test/TestUtilities/dataflow/FlowTestCommon.qll
Comment (MathiasVP) [reply cpp/ql/test/TestUtilities/dataflow/FlowTestCommon.qll:32]: Yep

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp
Comment (jketema) [inline cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp:597]: Is this really spurious, or is this just because we're also trying to track flow from uninitialised variables?

Comment (jketema) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp:597]: This is from commit 9533b98797aa7a3dddd58bb2df58032ec5f0d682

Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp:597]: Yeah, you're right. This is just because of the flow from the uninitialized variable. It's not really unitialized, though, since it's stack-allocated. So I think this might be something that should be fixed as part of https://github.com/github/codeql-c-team/issues/1409

File: cpp/ql/test/TestUtilities/dataflow/FlowTestCommon.qll
Comment (geoffw0) [reply cpp/ql/test/TestUtilities/dataflow/FlowTestCommon.qll:32]: Shouldn't this (and the other case like it) check the `File` matches as well?  We do sometimes have multiple files in a test so it seems like there's a risk of collision.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll
Comment (geoffw0) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll:111]: Why 2?

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (geoffw0) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:844]: I'm a bit worried about the performance of this, even with `bindingset[indirectionIndex]`, e.g. if `indirectionIndex` can take a known range of values and `t` is not bound.  You may well know better than I do here.  In any case, DCA should let us know if there's a serious problem...

**Review by geoffw0 — COMMENTED** [review]
In summary (I'm writing this mostly for my own benefit):
- we no longer flow from `op` with `*op` and `instr` to `*instr` in IR taint tracking.
- but we do in the default taint tracking layer instead, to preserve current query results.
- we do better flow through smart pointers, using the existing `PointerWrapper` models.
- arrays are now understood better by SSA, rather than being special cased in several places.
- fixed an issue with partial definitions (I understand why you did it, but not how it works).
- fixed some specific cases of indirections we didn't get right.
- fixed some performance issues.

Very thorough changes.  I appreciate the extra test cases and work on areas where there were result and performance regressions.

I have not reviewed the overall test changes (yet).

Definitely want to see a DCA run on this!

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/dataflow-ir-consistency.expected
Comment (geoffw0) [inline cpp/ql/test/library-tests/dataflow/dataflow-tests/dataflow-ir-consistency.expected:31]: What's the reason for all these new "PostUpdateNode should not be the target of local flow." consistency check results?  Are they serious?

File: cpp/ql/test/library-tests/dataflow/taint-tests/taint.expected
Comment (geoffw0) [inline cpp/ql/test/library-tests/dataflow/taint-tests/taint.expected:1]: The inline expectation for this should be corrected.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll:111]: Given a variable `void* p` we currently, we have a dataflow node for `p` and for `*p` with types `void*` and `void` respectively (we should probably have a better type for `*p` eventually, but that's left as future work). This means that we'll have flow in cases such as:
```cpp
int x = source(); // x is tainted
void* p = (void*)&x; // *p is tainted
sink(*(int*)p);
```
since there'll be flow from `source() -> x -> indirection of p -> *(int*)p`.

That's all well and good. Now, consider this slightly different program (that is equivalent to the situation added in https://github.com/github/codeql/pull/12316/commits/fddc54f4d26372212dc4cc147cb9ca6ef4cf0bff):
```cpp
char* p = indirect_source(); // *p is tainted
void* q = (void*)&p; // **q is tainted
sink(**(char**)q);
```
prior to this PR we wouldn't get flow in this case because we didn't have a dataflow node for `**q` (since `q` has type `void*` and thus we only created dataflow nodes for `q` and `*q`). This negatively impacted Samate quite a bit.

In the future I would we can do some pre analysis that decides how many dataflow nodes we need to allocate for a given void pointer (since, unlike with other pointer types, it's not clear how many possible dereferences people intend to do on it). But for now, we regain all the Samate results (without impacting performance as far as I can see) by just saying "we create a dataflow node for `q`, `*q` and `**q` for each void pointer `q`).

Does that make sense?

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/dataflow-ir-consistency.expected
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/dataflow-ir-consistency.expected:31]: It's never been a cause for concern before, no. The old IR dataflow library had many more inconsistency failures because of this, and we never saw an issue because of it.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll
Comment (geoffw0) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll:111]: I think I understand - we already had indirections 0 and 1, it's 2 because we want to support `**q` as that sort of thing is quite common.  And it could be `[2, 3]` instead to support `***q` as well but we have to stop somewhere.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll:111]: Yep, exactly.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:682]: I think I need a voice-over for this change. This is part of 9ad4e8138c0db74f08b36cfc429c1b3d313bd752.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:682]: Understandable! Consider this piece of code:
```cpp
int* p = indirect_source();
p[10] = 0;
sink(*p);
```
prior to this commit, we got flow in this example because we recognized that the definition `p[10] = 0` was a partial definition of `p`, and we'd use the shared SSA library's `uncertainWriteDefinitionInput` to find the previous definition (which in this case is `int* p = indirect_source()`), and thus we'd get flow from `int* p = indirect_source()` to `*p` because the SSA library concluded that there was flow from `p[10] = 0` to `sink(*p)` and `p[10] = 0` was partial. All good so far.

However, consider the failing testcase I added in https://github.com/github/codeql/commit/95222e37d67b0ee2a98ab3ac42ec67f83848f39f:
```cpp
intPointerSource(p);
p[10] = 0;
sink(*p); // $ MISSING: ast,ir
```
the old could would conclude that `p[10] = 0;` was a partial definition, and use `uncertainWriteDefinitionInput` to find the previous definition. However, `intPointerSource(p);` doesn't actually contain any SSA definitions. It does, however, contain an SSA _use_ of `*p`. So it turns out `uncertainWriteDefinitionInput` is the wrong tool for the job since it can only give us the previous definitions, but it won't have a result when the previous reference to the variable was a read instead of a def. So I swapped `uncertainWriteDefinitionInput` out with the right tool: `lastRefRedef`. That one gives us the (basic block, index) pair that last referenced the variable prior to a given definition, which is exactly what we need here (in this case the definition would be `p[10] = 0`, and the (basic block, index) pair would be the location of `p` in `intPointerSource(p)`).

Does that make sense?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:682]: I'd be happy to do sync on it if it's more helpful.

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:682]: Makes sense.

File: cpp/ql/src/Security/CWE/CWE-311/CleartextFileWrite.ql
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-311/CleartextFileWrite.ql:86]: This is cleaner than before the fix. :+1:

File: cpp/ql/test/query-tests/Security/CWE/CWE-311/semmle/tests/CleartextBufferWrite.expected
Comment (geoffw0) [inline cpp/ql/test/query-tests/Security/CWE/CWE-311/semmle/tests/CleartextBufferWrite.expected:22]: It looks like we've lost the result in `test2.cpp` line 110.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:1005]: I've gone back and forth to figure out exactly why this fix actually fixes all the missing results. The most high-level explanation (which is the one I currently understand myself 🤔) is that the dataflow node that receives the flow was not the one that was exposed as having a result for `asIndirectExpr()`. For example, consider this piece of code:
```cpp
void sink(char*);
void source(char*);

void test() {
  char buffer[256];
  source(buffer); // *buffer is tainted
  sink(buffer); // will have a array-to-pointer conversion
}
```
Dataflow will step from the output of `source` to the fully-converted `buffer` on the `sink(buffer)` line, but `asIndirectExpr()` would give back the _unconverted_ `buffer`.

Instead, this commit should hopefully ensure that only the fully-converted node is the node that has a result for `asIndirectExpr()`.

File: cpp/ql/src/Security/CWE/CWE-311/CleartextFileWrite.ql
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-311/CleartextFileWrite.ql:86]: Thanks! It should hopefully also be a bit faster since there'll be fewer sources and sinks considered by dataflow 🤞.

File: cpp/ql/test/query-tests/Security/CWE/CWE-311/semmle/tests/CleartextBufferWrite.expected
Comment (MathiasVP) [reply cpp/ql/test/query-tests/Security/CWE/CWE-311/semmle/tests/CleartextBufferWrite.expected:22]: Well spotted. This is actually an issue with our model for `gets`. I've opened a PR with a fix here: https://github.com/github/codeql/pull/12325

File: cpp/ql/test/library-tests/dataflow/taint-tests/taint.expected
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/taint.expected]: Oops. Missed that one. Fixed in https://github.com/github/codeql/pull/12316/commits/b36d4931f217f92f0090f98c46b03f54f02a0b26. Thanks!

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:844]: I think we should be good here. Most things should be covered by that first disjunct, so there won't be too many things hitting that second disjunct anyway. Maybe I should've actually put a `bindingset` annotation on `t` as well though (as it will actually be bound by all callers anyway).

File: cpp/ql/test/query-tests/Security/CWE/CWE-311/semmle/tests/CleartextBufferWrite.expected
Comment (MathiasVP) [reply cpp/ql/test/query-tests/Security/CWE/CWE-311/semmle/tests/CleartextBufferWrite.expected:24]: The above PR has been merged, and the fix is reflected in c69ae7f948154b2f9f07bbfb90d50d8f8debf2bc

**Comment by MathiasVP** [comment]
DCA and tests look good. I've verified that the following query result changes are good:
- `cpp/cleartext-transmission` on SAMATE
- `cpp/unbounded-write` on vim/vim
- `cpp/tainted-format-string` on vim/vim

Based on that sample I'm fairly confident in this PR now 🎉.

**Comment by jketema** [comment]
CI still shows a test failure for `Likely Bugs/Format/NonConstantFormat/NonConstantFormat.qlref`. I also see more failures on internal tests than I saw before, did you look at those?

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (geoffw0) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:844]: Looking at it today, I think I agree it's unlikely to be an issue.  Thanks!

**Comment by geoffw0** [comment]
LGTM apart from the test failures.  I also took a look at the DCA run - performance is fine - there are a lot of result changes!  Verified:
- `cpp/cleartext-storage-file` for git-linux
- `cpp/cleartext-storage-file` for systemd-linux
- `cpp/non-constant-format` for vim__vim
- `cpp/path-injection` for openjdk-jdk
- `cpp/tainted-format-string-through-global` for vim__vim (first two results only; it's fairly likely the various conditions in this extremely complex function make them safe in all actual cases, but the results are plausible).
- `cpp/uncontrolled-allocation-size` for git-linux (a lost result; I assume taint was coming in through the argument `refs` into `snapshot->refs`, so `snapshot->eof` and `snapshot->start` are indeed clean and we're happy to lose this result?)

**Comment by MathiasVP** [comment]
> CI still shows a test failure for `Likely Bugs/Format/NonConstantFormat/NonConstantFormat.qlref`.

Oh, yes. I searched for `FAILED(RESULT)` to correct CI failures, but didn't search for `FAILED(EXECUTION)`. I'll take a look 👍.

> I also see more failures on internal tests - I think - than I saw before, did you look at those?

I see 19 `FAILED(RESULT)` failures on CI, which I think is consistent with other PRs?

**Comment by jketema** [comment]
> I see 19 `FAILED(RESULT)` failures on CI, which I think is consistent with other PRs?

That seems the correct number indeed. I apparently mis-counted.

**Comment by MathiasVP** [comment]
I fixed CI in https://github.com/github/codeql/pull/12316/commits/85c7116e8f9e1bbd2a37e55f3ba3b73e74e6ec76 (it was actually a bad join that made a stress test fail. Good job, tests! 🎉), which then revealed some new FPs (see https://github.com/github/codeql/pull/12316/commits/1e5b235f4b27807b32742ded2b36ce2cf647593f), which I then fixed in https://github.com/github/codeql/pull/12316/commits/d93d22ba3ea8a74b84bab0590eaad4fd7929fcb9.

TLDR: This PR should be good to go now. I've started a new DCA to make sure my last change didn't do anything surprising.

**Review by geoffw0 — APPROVED** [review]
LGTM (but wait for DCA + @jketema 's :+1: as well before merging)

**Review by jketema — APPROVED** [review]
LGTM (if DCA is ok)

**Comment by MathiasVP** [comment]
DCA still looks good 🎉

**Pull request merged**
